subiquitycore/core.py

```diff
@@ -452,62 +443,66 @@ def next(loop, user_data):
                     ss.scripts = ss.scripts[1:]
                     if ss.scripts:
                         _run_script()
-            loop.set_alarm_in(delay, next)
+            self.loop.set_alarm_in(delay, next)
 
         ss.ns['c'] = c
         ss.ns['wait'] = wait
-        ss.ns['ui'] = self.common['ui']
+        ss.ns['ui'] = self.ui
 
-        self.common['loop'].set_alarm_in(0.06, _run_script)
+        self.loop.set_alarm_in(0.06, _run_script)
+
+    def unhandled_input(self, key):
+        if key == 'ctrl x':
+            self.signal.emit_signal('control-x-quit')
 
     def run(self):
-        if not hasattr(self, 'loop'):
-            if (self.common['opts'].run_on_serial and
-                    os.ttyname(0) != "/dev/ttysclp0"):
-                palette = self.STYLES_MONO
-                screen = urwid.raw_display.Screen()
-            else:
-                screen, palette = setup_screen(self.COLORS, self.STYLES)
+        if (self.opts.run_on_serial and
+                os.ttyname(0) != "/dev/ttysclp0"):
+            palette = self.STYLES_MONO
+            screen = urwid.raw_display.Screen()
+        else:
+            screen, palette = setup_screen(self.COLORS, self.STYLES)
 
-            self.common['loop'] = urwid.MainLoop(
-                self.common['ui'], palette=palette, screen=screen,
-                handle_mouse=False, pop_ups=True,
-                input_filter=self.common['input_filter'].filter)
+        self.loop = urwid.MainLoop(
```

Comment:
does this have a chance to re-init loop? i.e. if run() is called multiple times, it would create new loops now. Previously it didn't.

Maybe add a test for "if not self.loop" or some such?

I guess, run shouldn't be called multiple times anyway, and it's a bug to do so.

Reply:
Well before "hasattr(self, 'loop'):" was always false before: until this branch nothing ever set self.loop! So I just deleted the check. It could be "self.loop is not None" instead, but ehh. Calling run twice would be an awful bug.